### ISSUE TITLE
[Feature Request] Support for neovim apis

### ISSUE THREAD

**Issue by saccarosium** [lervag/wiki.vim#274]
Hi,
Would be very nice having support for [`vim.ui.select`](https://github.com/neovim/neovim/pull/15771) in this plugin. Thin would give the possibility to use almost every relevant neovim fuzzyfinder without adding support for every single one of them.

Is in the scope of the plugin having some neovim specific features, or you want to maintain a vim only features set?

**Comment by lervag** [comment]
> Hi, Would be very nice having support for [`vim.ui.select`](https://github.com/neovim/neovim/pull/15771) in this plugin.

Can you be more specific? Where; for which features?

**Comment by saccarosium** [comment]
Sorry, I should be more specific.

> Where; for which features?

As an equivalent to `WikiFzfPages`, `WikiFzfTags` etc. Currently if I want search functionalities over tags or notes in my wiki, I need to use one of the fuzzy finders listed in the thiry-party support in the README. I personally use a neovim reimplementation of fzf.vim in lua. Since is not compatible with the vimscript version I'm basecly stuck to keep installed the vimscript fzf plugin only to use this plugin.

By implementing this feature neovim users they can choose their fuzzy finder for those commands.

**Comment by lervag** [comment]
Sorry to bother you here, but could you go somewhat further. Let's say I did this, how would it look to you? What would be the wiki.vim interface you were using and how would it work? This is not fully clear to me, which is why I'm asking.

It seems you are asking for e.g. `:WikiPages` and for it to use `vim.ui.select` instead of `fzf#run(fzf#wrap(...))` for selecting. However, I've not really implemented any interfaces based on `vim.ui.select`, so having a detailed explanation is very helpful for me to understand how to provide this.

I want wiki.vim to work for both Vim and neovim, but I'm not afraid of relying on neovim specific features for things as long as I can also provide decent Vim fallbacks. It seems what you are asking is that I generalize e.g. the fzf integration and instead let fzf be a fallback mechanism.

**Comment by saccarosium** [comment]
## Steps

First we would create a `lua` folder at the root of the project, and put a new file, for example `ui_select.lua`, under `lua/wiki/ui_select.lua`.

We can create a function for getting files from the `g:wiki_root` like so:

```lua
local function get_notes(path)
    local files = {}
    local handle = vim.loop.fs_scandir(path)
    while handle do
        local name = vim.loop.fs_scandir_next(handle)
        if not name then
            break
        else
            table.insert(files, name)
        end
    end
    return files
end
```

> **Warning**
> Is probably a better idea to create a global function in vimscript that returns a dictionary, so we don't need to reimplement the same feature for every fuzzy finder (note that if you call a function that returns a dictionary in lua it will be convertent in a lua table)

After that we can implement a `vim.ui.select` command (is very poorly written only for demonstrations)

```lua
local root = vim.g.wiki_root
vim.api.nvim_create_user_command('WikiPages', function()
    -- takes a table for the listing
    vim.ui.select(get_notes(root), { prompt = "WikiPages> " }, function(choice)
        vim.cmd("edit " .. root .. "/" .. choice)
    end)
end, {})
```

At the end we can put something like this in `plugin/wiki.vim`:

```vim
if has('nvim')
  lua require("wiki.ui_select")
endif
```

It will source create the command only on neovim.

## Demo

> **Note**
> Don't focus on the code self but on the interface

### Default `ui.select`

https://user-images.githubusercontent.com/96259932/222953016-55a0274a-bf2a-4b2e-91f0-04f0bca6d415.mov

### [Fzf-lua](https://github.com/ibhagwan/fzf-lua)


https://user-images.githubusercontent.com/96259932/222953155-5a6dd823-7d70-47f9-b98d-5c68b51bd20b.mov

----

If you care about the feature I can make a PR and we could work there.

**Comment by lervag** [comment]
I think I was quite unclear, sorry. I did not mean to ask for the technical details on how to implement this. What I want to understand is how the feature should look to a user who do not care about any of the implementation details.

As far as I understand, what you are asking is for something like this:

* Change the commands `WikiFzf{Pages,Tags,Toc}` into simply `Wiki{Pages,Tags,Toc}`.
* Let the implementation on neovim use `vim.ui.select` (similar to how you've outlined it).
* Use something like the existing fzf functionality remain as a fallback for Vim.

Do I understand this correctly?

It seems a viable path forward would be to start by implement something like `WikiPages` in such a way that it works on both Vim and neovim, where Vim has a simple fallback selector.

----

Notice, though, that I already have a `wiki#ui#select` implementation within wiki.vim. I could also perhaps use `vim.ui.select` as an alternative for this one on neovim - that may need some work, though, as `vim.ui.select` relies on a different design principle.

Do you know if it is possible to use `vim.ui.select` in a _blocking_ manner that mimics something like this?

```lua
local candidates = [...]
local selected = nul
vim.ui.select(candidates, opts, function(choice)
  selected = choice
end)
-- ...
```

The reason I ask is because if this is possible, then I believe it is a trivial matter to make a `vim.ui.select` variant for neovim that will be used for a few features that uses `wiki#ui#select`. Then I could also perhaps expand the latter function and use it as a backend for `WikiPages` and similar.

**Comment by saccarosium** [comment]
> Change the commands WikiFzf{Pages,Tags,Toc} into simply Wiki{Pages,Tags,Toc}.

No, my idea is simply have `Wiki{Pages,Tags,Toc}` along side with `WikiFzf{Pages,Tags,Toc}` with the only difference being that `Wiki*` are loaded only on neovim. But having an universal `WikiPages` could be a better solution.

> Do you know if it is possible to use vim.ui.select in a blocking manner that mimics something like this?

For what I'm aware is blocking. Look at the demos for the default picker, see if it is the beaviour you want.

**Comment by lervag** [comment]
This issue will be resolved from #281.
